File: core/src/com/unciv/models/gamebasics/Building.java
package com.unciv.models.gamebasics;

import com.badlogic.gdx.utils.Predicate;
import com.unciv.logic.city.CityInfo;
import com.unciv.logic.civilization.CivilizationInfo;
import com.unciv.logic.city.CityConstructions;
import com.unciv.logic.city.IConstruction;
import com.unciv.logic.map.TileInfo;
import com.unciv.ui.ScienceVictoryScreen;
import com.unciv.ui.UnCivGame;
import com.unciv.ui.VictoryScreen;
import com.unciv.ui.pickerscreens.PolicyPickerScreen;
import com.unciv.models.linq.Linq;
import com.unciv.models.stats.FullStats;
import com.unciv.models.stats.NamedStats;

public class Building extends NamedStats implements IConstruction, ICivilopedia {
    public String description;
    public String requiredTech;

    public Technology GetRequiredTech() {
        return GameBasics.Technologies.get(requiredTech);
    }

    public int cost;
    public int maintainance = 0;
    public FullStats percentStatBonus;
    public FullStats specialistSlots;
    public FullStats greatPersonPoints;
    public int hurryCostModifier; // Extra cost percentage when purchasing
    public boolean isWonder = false;
    public String requiredBuilding;
    public String requiredBuildingInAllCities;
    public String requiredResource; // A strategic resource that will be consumed by this building
    public Linq requiredNearbyImprovedResources; // City can only be built if one of these resources is nearby - it must be improved!
    public String cannotBeBuiltWith;

    // Uniques
    public String providesFreeBuilding;
    public int freeTechs;
    public String unique; // for wonders which have individual functions that are totally unique


    /**
     * The bonus stats that a resource gets when this building is built
     */
    public FullStats resourceBonusStats;

    public FullStats getStats(Linq<String> adoptedPolicies){
        FullStats stats = this.clone();
        if (adoptedPolicies.contains("Organized Religion") &&
                new Linq<String>("Monument","Temple","Monastery").contains(name))
            stats.happiness+=1;

        if (adoptedPolicies.contains("Free Religion") &&
                new Linq<String>("Monument","Temple","Monastery").contains(name))
            stats.culture+=1;

        if (adoptedPolicies.contains("Entrepreneurship") &&
                new Linq<String>("Mint","Market","Bank","Stock Market").contains(name))
            stats.science+=1;

        if (adoptedPolicies.contains("Humanism") &&
                new Linq<String>("University","Observatory","Public School").contains(name))
            stats.science+=1;

        if (adoptedPolicies.contains("Theocracy") && name.equals("Temple"))
            percentStatBonus = new FullStats(){{gold=10;}};

        if (adoptedPolicies.contains("Free Thought") && name.equals("University"))
            percentStatBonus.science=50;

        if (adoptedPolicies.contains("Rationalism Complete") && !isWonder && stats.science>0)
            stats.gold+=1;

        if (adoptedPolicies.contains("Constitution") && isWonder)
            stats.culture+=2;

        return stats;
    }

    public String getDescription() {
        return getDescription(false,new Linq<String>());
    }

    public String getDescription(boolean forBuildingPickerScreen, Linq<String> adoptedPolicies) {
        FullStats stats = getStats(adoptedPolicies);
        StringBuilder stringBuilder = new StringBuilder();
        if (!forBuildingPickerScreen) stringBuilder.append("Cost: " + cost + "\r\n");
        if (isWonder) stringBuilder.append("Wonder\r\n");
        if (!forBuildingPickerScreen && requiredTech != null)
            stringBuilder.append("Requires " + requiredTech + " to be researched\r\n");
        if (!forBuildingPickerScreen && requiredBuilding != null)
            stringBuilder.append("Requires a " + requiredBuilding + " to be built in this city\r\n");
        if (!forBuildingPickerScreen && requiredBuildingInAllCities != null)
            stringBuilder.append("Requires a " + requiredBuildingInAllCities + " to be built in all cities\r\n");
        if (providesFreeBuilding != null)
            stringBuilder.append("Provides a free " + providesFreeBuilding + " in this city\r\n");
        if(description!=null) stringBuilder.append(description + "\r\n");
        if(!stats.toString().equals(""))
        stringBuilder.append(stats+"\r\n");
        if(this.percentStatBonus!=null){
            if(this.percentStatBonus.production!=0) stringBuilder.append("+"+(int)this.percentStatBonus.production+"% production\r\n");
            if(this.percentStatBonus.gold!=0) stringBuilder.append("+"+(int)this.percentStatBonus.gold+"% gold\r\n");
            if(this.percentStatBonus.science!=0) stringBuilder.append("+"+(int)this.percentStatBonus.science+"% science\r\n");
            if(this.percentStatBonus.food!=0) stringBuilder.append("+"+(int)this.percentStatBonus.food+"% food\r\n");
            if(this.percentStatBonus.culture!=0) stringBuilder.append("+"+(int)this.percentStatBonus.culture+"% culture\r\n");
        }
        if(this.greatPersonPoints!=null){
            if(this.greatPersonPoints.production!=0) stringBuilder.append("+"+(int)this.greatPersonPoints.production+" Great Engineer points\r\n");
            if(this.greatPersonPoints.gold!=0) stringBuilder.append("+"+(int)this.greatPersonPoints.gold+" Great Merchant points\r\n");
            if(this.greatPersonPoints.science!=0) stringBuilder.append("+"+(int)this.greatPersonPoints.science+" Great Scientist points\r\n");
            if(this.greatPersonPoints.culture!=0) stringBuilder.append("+"+(int)this.greatPersonPoints.culture+" Great Artist points\r\n");
        }
        if(resourceBonusStats!=null){
            String resources = StringUtils.join(",",GameBasics.TileResources.linqValues().where(new Predicate<TileResource>() {
                @Override
                public boolean evaluate(TileResource arg0) {
                    return name.equals(arg0.building);
                }
            }).select(new Linq.Func<TileResource, String>() {
                @Override
                public String GetBy(TileResource arg0) {
                    return arg0.name;
                }
            })) ;
            stringBuilder.append(resources+" provide "+resourceBonusStats+" \r\n");
        }
        if (maintainance != 0)
            stringBuilder.append("Maintainance cost: " + maintainance + " gold\r\n");
        return stringBuilder.toString();
    }

    @Override
    public int getProductionCost(Linq<String> adoptedPolicies) {
        if(!isWonder && culture!=0 && adoptedPolicies.contains("Piety"))
            return (int) (cost*0.85);
        return cost;
    }

    public int getGoldCost(Linq<String> adoptedPolicies) {
        double cost = Math.pow(30 * getProductionCost(adoptedPolicies),0.75) * (1 + hurryCostModifier/100);
        if(adoptedPolicies.contains("Mercantilism")) cost*=0.75;
        if(adoptedPolicies.contains("Patronage")) cost*=0.5;
        return (int)( cost / 10 ) * 10;
    }
    
    public boolean isBuildable(CityConstructions construction){
        CivilizationInfo civInfo = construction.cityInfo.civInfo;
        if(construction.isBuilt(name)) return false;
        if (requiredTech != null && !civInfo.tech.isResearched(requiredTech)) return false;
        if (isWonder && civInfo.cities
                .any(new Predicate<CityInfo>() {
                    @Override
                    public boolean evaluate(CityInfo arg0) {
                        CityConstructions CB = arg0.cityConstructions;
                        return CB.isBuilding(name) || CB.isBuilt(name);
                    }
                }) ) return false;
        if (requiredBuilding != null && !construction.isBuilt(requiredBuilding)) return false;
        if (requiredBuildingInAllCities != null ||
                civInfo.cities.any(new Predicate<CityInfo>() {
                    @Override
                    public boolean evaluate(CityInfo arg0) {
                        return arg0.cityConstructions.isBuilt(requiredBuildingInAllCities);
                    }
                }) ) return false;
        if(cannotBeBuiltWith != null && construction.isBuilt(cannotBeBuiltWith)) return false;
        if("MustBeNextToDesert".equals(unique) &&
                !civInfo.gameInfo.tileMap.getTilesInDistance(construction.cityInfo.cityLocation,1).any(new Predicate<TileInfo>() {
                    @Override
                    public boolean evaluate(TileInfo arg0) {
                        return arg0.baseTerrain.equals("Desert");
                    }
                }))
            return false;
        if(requiredResource!=null &&
                !civInfo.getCivResources().keySet().contains(GameBasics.TileResources.get(requiredResource)))
            return false; // Only checks if exists, doesn't check amount - todo


        if(requiredNearbyImprovedResources!=null) {
            boolean containsResourceWithImprovement = construction.cityInfo.getTilesInRange()
                    .any(new Predicate<TileInfo>() {
                        @Override
                        public boolean evaluate(TileInfo tile) {
                            return tile.resource != null
                                    && requiredNearbyImprovedResources.contains(tile.resource)
                                    && tile.getTileResource().improvement.equals(tile.improvement);
                        }
                    });
            if(!containsResourceWithImprovement) return false;
        }

        if("SpaceshipPart".equals(unique)){
            if(!civInfo.getBuildingUniques().contains("ApolloProgram")) return false;
            if(civInfo.scienceVictory.unconstructedParts().get(name) ==0) return false; // Don't need to build any more of these!
        }
        return true;
    }

    @Override
    public void postBuildEvent(CityConstructions constructions) {
        CivilizationInfo civInfo = constructions.cityInfo.civInfo;
        if("ApolloProgram".equals(unique)) UnCivGame.Current.setScreen(new ScienceVictoryScreen(civInfo));
        if("SpaceshipPart".equals(unique)) {
            civInfo.scienceVictory.currentParts.add(name, 1);
            UnCivGame.Current.setScreen(new ScienceVictoryScreen(civInfo));
            if(civInfo.scienceVictory.unconstructedParts().isEmpty())
                UnCivGame.Current.setScreen(new VictoryScreen());
            return;
        }
        constructions.builtBuildings.add(name);
        if (providesFreeBuilding != null && !constructions.builtBuildings.contains(providesFreeBuilding))
            constructions.builtBuildings.add(providesFreeBuilding);
        if (freeTechs != 0) civInfo.tech.freeTechs += freeTechs;
        if("EmpireEntersGoldenAge".equals(unique)) civInfo.goldenAges.enterGoldenAge();
        if("FreeGreatArtistAppears".equals(unique)) civInfo.addGreatPerson("Great Artist");
        if("WorkerConstruction".equals(unique)){
            civInfo.placeUnitNearTile(constructions.cityInfo.cityLocation,"Worker");
            civInfo.placeUnitNearTile(constructions.cityInfo.cityLocation,"Worker");
        }
        if("FreeSocialPolicy".equals(unique)){
            civInfo.policies.freePolicies++;
            UnCivGame.Current.setScreen(new PolicyPickerScreen(civInfo));
        }
    }
}

File: core/src/com/unciv/models/gamebasics/Building.kt
package com.unciv.models.gamebasics

import com.unciv.logic.city.CityConstructions
import com.unciv.logic.city.IConstruction
import com.unciv.logic.civilization.CivilizationInfo
import com.unciv.models.gamebasics.tech.Technology
import com.unciv.models.stats.NamedStats
import com.unciv.models.stats.Stats

class Building : NamedStats(), IConstruction{

    override val description: String
        get() = getDescription(false, null)

    var requiredTech: String? = null

    var cost: Int = 0
    var maintenance = 0
    var percentStatBonus: Stats? = null
    var specialistSlots: Stats? = null
    var greatPersonPoints: Stats? = null
    /** Extra cost percentage when purchasing */
    var hurryCostModifier: Int = 0
    var isWonder = false
    var requiredBuilding: String? = null
    var requiredBuildingInAllCities: String? = null
    /** A strategic resource that will be consumed by this building */
    var requiredResource: String? = null
    /** City can only be built if one of these resources is nearby - it must be improved! */
    var requiredNearbyImprovedResources: List<String>? = null
    var cannotBeBuiltWith: String? = null
    var cityStrength=0
    var cityHealth=0
    var xpForNewUnits=0
    var replaces:String?=null
    var uniqueTo:String?=null

    // Uniques
    var providesFreeBuilding: String? = null
    var freeTechs: Int = 0
    var uniques = ArrayList<String>()


    /**
     * The bonus stats that a resource gets when this building is built
     */
    var resourceBonusStats: Stats? = null

    fun getRequiredTech(): Technology = GameBasics.Technologies[requiredTech]!!

    fun getShortDescription(): String { // should fit in one line
        val infoList= mutableListOf<String>()
        val str = getStats(null).toString()
        if(str.isNotEmpty()) infoList += str
        if(percentStatBonus!=null){
            for(stat in percentStatBonus!!.toHashMap())
                if(stat.value!=0f) infoList+="+${stat.value.toInt()}% ${stat.key.toString().tr()}"
        }
        val improvedResources = GameBasics.TileResources.values.filter { it.building==name }.map { it.name.tr() }
        if(improvedResources.isNotEmpty()){
            // buildings that improve resources
            infoList += improvedResources.joinToString()+ " {provide} ".tr()+ resourceBonusStats.toString()
        }
        if(requiredNearbyImprovedResources!=null)
            infoList += ("Requires worked ["+requiredNearbyImprovedResources!!.joinToString("/"){it.tr()}+"] near city").tr()
        if(uniques.isNotEmpty()) infoList += uniques.joinToString { it.tr() }
        if(cityStrength!=0) infoList+="{City strength} +".tr()+cityStrength
        if(cityHealth!=0) infoList+="{City health} +".tr()+cityHealth
        if(xpForNewUnits!=0) infoList+= "+$xpForNewUnits {XP for new units}".tr()
        return infoList.joinToString()
    }

    fun getDescription(forBuildingPickerScreen: Boolean, civInfo: CivilizationInfo?): String {
        val stats = getStats(civInfo)
        val stringBuilder = StringBuilder()
        if(uniqueTo!=null) stringBuilder.appendln("Unique to [$uniqueTo], replaces [$replaces]".tr())
        if (!forBuildingPickerScreen) stringBuilder.appendln("{Cost}: $cost".tr())
        if (isWonder) stringBuilder.appendln("Wonder".tr())
        if (!forBuildingPickerScreen && requiredTech != null)
            stringBuilder.appendln("Required tech: [$requiredTech]".tr())
        if (!forBuildingPickerScreen && requiredBuilding != null)
            stringBuilder.appendln("Requires [$requiredBuilding] to be built in the city".tr())
        if (!forBuildingPickerScreen && requiredBuildingInAllCities != null)
            stringBuilder.appendln("Requires [$requiredBuildingInAllCities] to be built in all cities".tr())
        if (providesFreeBuilding != null)
            stringBuilder.appendln("Provides a free [$providesFreeBuilding] in the city".tr())
        if(uniques.isNotEmpty()) stringBuilder.appendln(uniques.asSequence().map { it.tr() }.joinToString("\n"))
        if (stats.toString() != "")
            stringBuilder.appendln(stats)
        if (this.percentStatBonus != null) {
            if (this.percentStatBonus!!.production != 0f) stringBuilder.append("+" + this.percentStatBonus!!.production.toInt() + "% {Production}\n".tr())
            if (this.percentStatBonus!!.gold != 0f) stringBuilder.append("+" + this.percentStatBonus!!.gold.toInt() + "% {Gold}\n".tr())
            if (this.percentStatBonus!!.science != 0f) stringBuilder.append("+" + this.percentStatBonus!!.science.toInt() + "% {Science}\r\n".tr())
            if (this.percentStatBonus!!.food != 0f) stringBuilder.append("+" + this.percentStatBonus!!.food.toInt() + "% {Food}\n".tr())
            if (this.percentStatBonus!!.culture != 0f) stringBuilder.append("+" + this.percentStatBonus!!.culture.toInt() + "% {Culture}\r\n".tr())
        }
        if (this.greatPersonPoints != null) {
            val gpp = this.greatPersonPoints!!
            if (gpp.production != 0f) stringBuilder.appendln("+" + gpp.production.toInt()+" " + "[Great Engineer] points".tr())
            if (gpp.gold != 0f) stringBuilder.appendln("+" + gpp.gold.toInt() + " "+"[Great Merchant] points".tr())
            if (gpp.science != 0f) stringBuilder.appendln("+" + gpp.science.toInt() + " "+"[Great Scientist] points".tr())
            if (gpp.culture != 0f) stringBuilder.appendln("+" + gpp.culture.toInt() + " "+"[Great Artist] points".tr())
        }
        if (resourceBonusStats != null) {
            val resources = GameBasics.TileResources.values.filter { name == it.building }.joinToString { it.name.tr() }
            stringBuilder.appendln("$resources {provide} $resourceBonusStats".tr())
        }

        if(requiredNearbyImprovedResources!=null)
            stringBuilder.appendln(("Requires worked ["+requiredNearbyImprovedResources!!.joinToString("/"){it.tr()}+"] near city").tr())

        if(cityStrength!=0) stringBuilder.appendln("{City strength} +".tr() + cityStrength)
        if(cityHealth!=0) stringBuilder.appendln("{City health} +".tr() + cityHealth)
        if(xpForNewUnits!=0) stringBuilder.appendln("+$xpForNewUnits {XP for new units}".tr())
        if (maintenance != 0)
            stringBuilder.appendln("{Maintenance cost}: $maintenance {Gold}".tr())
        return stringBuilder.toString().trim()
    }

    val cultureBuildings = hashSetOf("Monument", "Temple", "Monastery")

    fun getStats(civInfo: CivilizationInfo?): Stats {
        val stats = this.clone()
        if(civInfo != null) {
            val adoptedPolicies = civInfo.policies.adoptedPolicies
            if (adoptedPolicies.contains("Organized Religion") && cultureBuildings.contains(name))
                stats.happiness += 1

            if (adoptedPolicies.contains("Free Religion") && cultureBuildings.contains(name))
                stats.culture += 1f

            if (adoptedPolicies.contains("Entrepreneurship") && hashSetOf("Mint", "Market", "Bank", "Stock Market").contains(name))
                stats.science += 1f

            if (adoptedPolicies.contains("Humanism") && hashSetOf("University", "Observatory", "Public School").contains(name))
                stats.happiness += 1f

            if (adoptedPolicies.contains("Theocracy") && name == "Temple")
                percentStatBonus = Stats().apply { gold = 10f }

            if (adoptedPolicies.contains("Free Thought") && name == "University")
                percentStatBonus!!.science = 50f

            if (adoptedPolicies.contains("Rationalism Complete") && !isWonder && stats.science > 0)
                stats.gold += 1f

            if (adoptedPolicies.contains("Constitution") && isWonder)
                stats.culture += 2f

            if (adoptedPolicies.contains("Autocracy Complete") && cityStrength > 0)
                stats.happiness += 1

            if (name == "Castle" && civInfo.getBuildingUniques().contains("+1 happiness, +2 culture and +3 gold from every Castle")){
                stats.happiness+=1
                stats.culture+=2
                stats.gold+=3
            }
        }
        return stats
    }

    override fun canBePurchased(): Boolean {
        return !isWonder
    }

    override fun getProductionCost(adoptedPolicies: HashSet<String>): Int {
        return if (!isWonder && culture != 0f && adoptedPolicies.contains("Piety")) (cost * 0.85).toInt()
        else cost
    }

    override fun getGoldCost(civInfo: CivilizationInfo, baseCost: Boolean): Int {
        // https://forums.civfanatics.com/threads/rush-buying-formula.393892/
        var cost: Double
        if (baseCost) {
            cost = Math.pow((30 * getProductionCost(hashSetOf())).toDouble(), 0.75) * (1 + hurryCostModifier / 100)
        } else {
            cost = Math.pow((30 * getProductionCost(civInfo.policies.adoptedPolicies)).toDouble(), 0.75) * (1 + hurryCostModifier / 100)
            if (civInfo.policies.adoptedPolicies.contains("Mercantilism")) cost *= 0.75
            if (civInfo.getBuildingUniques().contains("-15% to purchasing items in cities")) cost *= 0.85
            if (civInfo.policies.adoptedPolicies.contains("Patronage")
                    && listOf("Monument", "Temple", "Opera House", "Museum").contains(name)) cost *= 0.5
        }

        return (cost / 10).toInt() * 10
    }


    override fun shouldBeDisplayed(construction: CityConstructions): Boolean {
        val rejectionReason = getRejectionReason(construction)
        return rejectionReason==""
                || rejectionReason.startsWith("Requires")
                || rejectionReason == "Wonder is being built elsewhere"
    }

    fun getRejectionReason(construction: CityConstructions):String{
        if (construction.isBuilt(name)) return "Already built"

        val civInfo = construction.cityInfo.civInfo
        if (uniqueTo!=null && uniqueTo!=civInfo.civName) return "Unique to $uniqueTo"
        if (GameBasics.Buildings.values.any { it.uniqueTo==civInfo.civName && it.replaces==name }) return "Our unique building replaces this"
        if (requiredTech != null && !civInfo.tech.isResearched(requiredTech!!)) return "$requiredTech not researched"

        // Regular wonders
        if (isWonder && requiredBuildingInAllCities==null){
            if(civInfo.gameInfo.civilizations.flatMap { it.cities }
                            .any {it.cityConstructions.isBuilt(name)})
                return "Wonder is already built"

            if(civInfo.cities.any { it!=construction.cityInfo && it.cityConstructions.isBeingConstructed(name) })
                return "Wonder is being built elsewhere"
        }


        // National wonders
        if(requiredBuildingInAllCities!=null) {
            if (civInfo.cities.any { !it.cityConstructions.containsBuildingOrEquivalent(requiredBuildingInAllCities!!) })
                return "Requires a $requiredBuildingInAllCities in all cities"

            if (civInfo.cities.any {it.cityConstructions.isBuilt(name) })
                return "Wonder is already built"
            if (civInfo.cities.any {it.cityConstructions.isBeingConstructed(name) })
                return "Wonder is being built elsewhere"
        }

        if (requiredBuilding != null && !construction.containsBuildingOrEquivalent(requiredBuilding!!))
            return "Requires a $requiredBuilding in this city"
        if (cannotBeBuiltWith != null && construction.isBuilt(cannotBeBuiltWith!!))
            return "Cannot be built with $cannotBeBuiltWith"

        if ("Must be next to desert" in uniques
                && !construction.cityInfo.getCenterTile().getTilesInDistance(1).any { it.baseTerrain == "Desert" })
            return "Must be next to desert"

        if ("Must be next to mountain" in uniques
                && !construction.cityInfo.getCenterTile().getTilesInDistance(1).any { it.baseTerrain == "Mountain" })
            return "Must be next to mountain"

        if("Can only be built in coastal cities" in uniques
                && construction.cityInfo.getCenterTile().neighbors.none { it.baseTerrain=="Coast" })
            return "Can only be built in coastal cities"

        if (requiredResource != null && !civInfo.hasResource(requiredResource!!))
            return "Requires $requiredResource"

        if (requiredNearbyImprovedResources != null) {
            val containsResourceWithImprovement = construction.cityInfo.getTilesInRange()
                    .any {
                        it.resource != null
                                && requiredNearbyImprovedResources!!.contains(it.resource!!)
                                && it.getTileResource().improvement == it.improvement
                                && it.getOwner() == civInfo
                    }
            if (!containsResourceWithImprovement) return "Nearby $requiredNearbyImprovedResources required"
        }

        if ("Spaceship part" in uniques) {
            if (!civInfo.getBuildingUniques().contains("Enables construction of Spaceship parts")) return "Apollo project not built!"
            if (civInfo.victoryManager.unconstructedSpaceshipParts()[name] == 0) return "Don't need to build any more of these!"
        }
        return ""
    }

    override fun isBuildable(construction: CityConstructions): Boolean {
        return getRejectionReason(construction)==""
    }

    override fun postBuildEvent(construction: CityConstructions) {
        val civInfo = construction.cityInfo.civInfo

        if ("Spaceship part" in uniques) {
            civInfo.victoryManager.currentsSpaceshipParts.add(name, 1)
            return
        }
        construction.addBuilding(name)

        if (providesFreeBuilding != null && !construction.containsBuildingOrEquivalent(providesFreeBuilding!!)) {
            var buildingToAdd = providesFreeBuilding!!

            for(building in GameBasics.Buildings.values)
                if(building.replaces == buildingToAdd && building.uniqueTo==civInfo.civName)
                    buildingToAdd = building.name

            construction.addBuilding(buildingToAdd)
        }

        when {
            "Empire enters golden age" in uniques -> civInfo.goldenAges.enterGoldenAge()
            "Free Great Artist Appears" in uniques -> civInfo.addGreatPerson("Great Artist", construction.cityInfo)
            "Free great scientist appears" in uniques -> civInfo.addGreatPerson("Great Scientist", construction.cityInfo)
            "2 free great scientists appear" in uniques -> {
                civInfo.addGreatPerson("Great Scientist", construction.cityInfo)
                civInfo.addGreatPerson("Great Scientist", construction.cityInfo)
            }
            "Provides 2 free workers" in uniques -> {
                civInfo.placeUnitNearTile(construction.cityInfo.location, "Worker")
                civInfo.placeUnitNearTile(construction.cityInfo.location, "Worker")
            }
            "Free Social Policy" in uniques -> civInfo.policies.freePolicies++
            "Free Great Person" in uniques -> {
                if (civInfo.isPlayerCivilization()) civInfo.greatPeople.freeGreatPeople++
                else civInfo.addGreatPerson(GameBasics.Units.keys.filter { it.startsWith("Great") }.random())
            }
            "+1 population in each city" in uniques -> {
                for(city in civInfo.cities){
                    city.population.population += 1
                    city.population.autoAssignPopulation()
                }
            }
            "Enemy land units must spend 1 extra movement point when inside your territory (obsolete upon Dynamite)" in uniques ->
                civInfo.updateHasActiveGreatWall()
        }

        if (freeTechs != 0) civInfo.tech.freeTechs += freeTechs
    }
}